### ISSUE TITLE
Cant build on Ubuntu 

### ISSUE THREAD

**Issue by AffectedArc07** [hoodie/notify-rust#177]
OS: `OS: Ubuntu 20.04.4 LTS on Windows 10 x86_64` (Yes I am running under WSL)

I have `notify-rust` imported into another project as
```toml
notify-rust = { version = "4.6.0", optional = true }
```

However, upon building I get the following error:
```sh
aa07@AA07-C:/mnt/c/Users/AffectedArc07/Documents/GitHub/Parastuff/rust-g/paradise-rust-g$ cargo build --release --all-features --target=i686-unknown-linux-gnu
   Compiling rust-g v1.2.0-P2 (/mnt/c/Users/AffectedArc07/Documents/GitHub/Parastuff/rust-g/paradise-rust-g)
   Compiling notify-rust v4.6.0
error[E0432]: unresolved import `zbus::MatchRule`
 --> /home/aa07/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:2:34
  |
2 | use zbus::{blocking::Connection, MatchRule};
  |                                  ^^^^^^^^^ no `MatchRule` in the root

error[E0599]: no method named `add_match_rule` found for struct `zbus::blocking::fdo::DBusProxy<'_>` in the current scope
   --> /home/aa07/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:152:11
    |
152 |     proxy.add_match_rule(action_signal_rule).unwrap();
    |           ^^^^^^^^^^^^^^ help: there is a method with a similar name: `add_match`

error[E0599]: no method named `add_match_rule` found for struct `zbus::blocking::fdo::DBusProxy<'_>` in the current scope
   --> /home/aa07/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:161:11
    |
161 |     proxy.add_match_rule(close_signal_rule).unwrap();
    |           ^^^^^^^^^^^^^^ help: there is a method with a similar name: `add_match`

Some errors have detailed explanations: E0432, E0599.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `notify-rust` due to 3 previous errors
```

Is this crate not compatible with `i686-unknown-linux-gnu`? Am I missing an extra compile dependency?

**Comment by hoodie** [comment]
Could be that the target configuration gets confused because you’re building on WSL.

**Comment by AffectedArc07** [comment]
Occurs on a real install as well
![image](https://user-images.githubusercontent.com/25063394/212021660-9ac9c174-cc5c-45fd-b9cd-d3e0388702e8.png)

`OS: Ubuntu 20.04.5 LTS x86_64`

**Comment by hoodie** [comment]
also when you leave out `--all-features`?

**Comment by AffectedArc07** [comment]
> also when you leave out `--all-features`?

Let me modify the library to build the feature that needs `notify-rust` by default and ill get back to you.

**Comment by AffectedArc07** [comment]
Update - nada

```
root@AA07-S-UBUNTU:~/code/para-rust-g/paradise-rust-g# cargo build --release --target=i686-unknown-linux-gnu
   Compiling hyper-rustls v0.23.0
   Compiling git2 v0.14.4
   Compiling notify-rust v4.6.0
error[E0432]: unresolved import `zbus::MatchRule`
 --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:2:34
  |
2 | use zbus::{blocking::Connection, MatchRule};
  |                                  ^^^^^^^^^ no `MatchRule` in the root

   Compiling reqwest v0.11.11
error[E0599]: no method named `add_match_rule` found for struct `zbus::blocking::fdo::DBusProxy<'_>` in the current scope
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:152:11
    |
152 |     proxy.add_match_rule(action_signal_rule).unwrap();
    |           ^^^^^^^^^^^^^^ help: there is a method with a similar name: `add_match`

error[E0599]: no method named `add_match_rule` found for struct `zbus::blocking::fdo::DBusProxy<'_>` in the current scope
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:161:11
    |
161 |     proxy.add_match_rule(close_signal_rule).unwrap();
    |           ^^^^^^^^^^^^^^ help: there is a method with a similar name: `add_match`

Some errors have detailed explanations: E0432, E0599.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `notify-rust` due to 3 previous errors
```

**Comment by hoodie** [comment]
I don't really have a clue how to explain this at this point. I can however assume that building it on WSL won't do what you expect, because the linux binary will not talk to the windows api. it will try to send a notification to dbus, which you probably don't run on windows unless you have a linux desktop environment as well.
as for why some parts of the zbus library aren't there: 🤷

**Comment by AffectedArc07** [comment]
> I don't really have a clue how to explain this at this point. I can however assume that building it on WSL won't do what you expect, because the linux binary will not talk to the windows api. it will try to send a notification to dbus, which you probably don't run on windows unless you have a linux desktop environment as well. as for why some parts of the zbus library aren't there: 🤷

All builds after the first were performed on a real baremetal ubuntu server install that I use purely as a compile host. I should probably give some background to what I am achieving here to simplify things.

I maintain an open source game project (SS13), and while development happens entirely on windows (client only runs on windows), you can run the dedicated server on windows and linux. This library is used in an external rust DLL (or .so on linux) that is called from the main game, and my intention is for notification support while running on windows, but I dont need any support in the slightest on linux, other than "it compiles", hence me trying to compile it on a headless OS. 

Do you know of a `libzbus-dev` or similarly named package that has the developer stuff to compile with without the full support?

**Comment by hoodie** [comment]
Thanks for the clarification.
You could try to disable the default `"z"` feature and turn on the `"d"` feature, which switches the underlying dbus library from the cool pure-rust zbus, to the good'ole libdbus-somethingsomething. That should simplify your builds, but you need libdbus in your pkg-conf somewhere. If you don't need the feature at all on linux you might wanna try out something like `#[cfg(target_os = "windows")]` or similar, to not pull in the lib at all unless you're on windows.

**Comment by AffectedArc07** [comment]
> Thanks for the clarification. You could try to disable the default `"z"` feature and turn on the `"d"` feature, which switches the underlying dbus library from the cool pure-rust zbus, to the good'ole libdbus-somethingsomething. That should simplify your builds, but you need libdbus in your pkg-conf somewhere. If you don't need the feature at all on linux you might wanna try out something like `#[cfg(target_os = "windows")]` or similar, to not pull in the lib (notify-rust) at all unless you're on windows.

```
root@AA07-S-UBUNTU:~/code/para-rust-g/paradise-rust-g# cargo build --release --target=i686-unknown-linux-gnu
   Compiling notify-rust v4.6.0
error[E0432]: unresolved import `zbus::MatchRule`
 --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:2:34
  |
2 | use zbus::{blocking::Connection, MatchRule};
  |                                  ^^^^^^^^^ no `MatchRule` in the root
  |
help: consider importing this struct instead
  |
2 | use zbus::{blocking::Connection, dbus::message::MatchRule;
  |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0599]: no method named `into_i32` found for enum `timeout::Timeout` in the current scope
  --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:81:38
   |
81 |                 notification.timeout.into_i32(),
   |                                      ^^^^^^^^ method not found in `timeout::Timeout`
   |
  ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/timeout.rs:3:1
   |
3  | pub enum Timeout {
   | ---------------- method `into_i32` not found for this enum

error[E0599]: no method named `add_match_rule` found for struct `zbus::blocking::fdo::DBusProxy<'_>` in the current scope
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:152:11
    |
152 |     proxy.add_match_rule(action_signal_rule).unwrap();
    |           ^^^^^^^^^^^^^^ help: there is a method with a similar name: `add_match`

error[E0599]: no method named `add_match_rule` found for struct `zbus::blocking::fdo::DBusProxy<'_>` in the current scope
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/notify-rust-4.6.0/src/xdg/zbus_rs.rs:161:11
    |
161 |     proxy.add_match_rule(close_signal_rule).unwrap();
    |           ^^^^^^^^^^^^^^ help: there is a method with a similar name: `add_match`
```

When using
`notify-rust = { version = "4.6.0", optional = true, features = ["d"] }`

Do tell me if I scuffed something up with this

---

As for making this crate only load in if windows is targeted, that is an option and something I could do, will let you know how that goes.

**Comment by hoodie** [comment]
`z` is a default features, you have to build it with `no-default-features = true`

**Comment by AffectedArc07** [comment]
> `z` is a default features, you have to build it with `no-default-features = true`
> 
> https://github.com/hoodie/notify-rust#d

This worked, thank you

**Comment by Miniontoby** [comment]
Hey guys, 

I can also not build on ubuntu (using github actions)

I already created an issue over at https://github.com/notify-rs/notify/issues/469 but they said that that repo was no the correct one, but this one is. 

I have tried the `notify-rust = { version = "4", no-default-features = true, features = ["d"] }`

Or did I implement it incorrectly?

**Comment by hoodie** [comment]
I think this issue occurs when you try to build with both the `z` and `d` features.
`d` explicitly turns off this implementation and `z` includes code that needs it. These two features were not intended to be used together, but are also not explicitly mutually exclusive. I think I ought to fix that one day.

As for your concrete issue: I assume that you didn't turn off the `z` default-feature correctly somehow, maybe a typo or something?

**Comment by Miniontoby** [comment]
> I think this issue occurs when you try to build with both the `z` and `d` features. `d` explicitly turns off this implementation and `z` includes code that needs it. These two features were not intended to be used together, but are also not explicitly mutually exclusive. I think I ought to fix that one day.
> 
> As for your concrete issue: I assume that you didn't turn off the `z` default-feature correctly somehow, maybe a typo or something?

Possibly yes, but idk... Btw when I just don't have the `features = ["d"]` it also accures!

**Comment by hoodie** [comment]
then the feature seems to be enabled some other way. Do you pull in notify-rust directly or via a different dependency? I think this thing might fix the issue:
https://github.com/hoodie/notify-rust/pull/180

**Comment by Miniontoby** [comment]
> then the feature seems to be enabled some other way. Do you pull in notify-rust directly or via a different dependency? I think this thing might fix the issue: #180

I just use cargo! I don't pull it or so. Btw I just was looking on the internet and it seems like `no-default-features = true` is incorrect, but `default-features = false` is the correct maybe... So I will just try that and see what happens

**Comment by Miniontoby** [comment]
> > then the feature seems to be enabled some other way. Do you pull in notify-rust directly or via a different dependency? I think this thing might fix the issue: #180
> 
> I just use cargo! I don't pull it or so. Btw I just was looking on the internet and it seems like `no-default-features = true` is incorrect, but `default-features = false` is the correct maybe... So I will just try that and see what happens

That fixed it (even without `features = ["d"]`)
